src/protocols/gamma_control.rs

```diff
@@ -0,0 +1,237 @@
+                let output_id = output.id();
+                let gamma_size = state.get_gamma_size(&output);
+                let previous_gamma_ramp = state.get_gamma(&output);
+                let mut got_gamma_map = true;
+                if previous_gamma_ramp.is_none() {
+                    got_gamma_map = false;
+                }
+
+                let zwlr_gamma_control = data_init.init(
+                    id,
+                    GammaControlState {
+                        gamma_size: gamma_size.clone(),
+                        previous_gamma_ramp,
+                        output,
+                    },
+                );
+                if !got_gamma_map {
+                    zwlr_gamma_control.failed();
+                    error!("Failed to get current gamma ramp");
+                    return;
+                }
+
+                if state
+                    .gamma_control_manager_state()
+                    .gamma_controls
+                    .contains_key(&output_id)
+                    || gamma_size.is_none()
+                {
+                    zwlr_gamma_control.failed();
+                    return;
+                }
+
+                if gamma_size.is_none() {
+                    zwlr_gamma_control.failed();
+                    error!("gamma size not set");
+                    return;
+                }
+
+                zwlr_gamma_control.gamma_size(gamma_size.unwrap());
+                state
+                    .gamma_control_manager_state()
+                    .gamma_controls
+                    .insert(output_id, zwlr_gamma_control);
+            }
+            zwlr_gamma_control_manager_v1::Request::Destroy => (),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<D> Dispatch<ZwlrGammaControlV1, GammaControlState, D> for GammaControlManagerState
+where
+    D: Dispatch<ZwlrGammaControlV1, GammaControlState>,
+    D: GammaControlHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        resource: &ZwlrGammaControlV1,
+        request: <ZwlrGammaControlV1 as Resource>::Request,
+        data: &GammaControlState,
+        _dhandle: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            zwlr_gamma_control_v1::Request::SetGamma { fd } => {
+                info!("Setting gamma for output {:?}", data.output);
+                let buf = &mut Vec::new();
+                if File::from(fd).read_to_end(buf).is_err() {
+                    error!("Failed to read gamma data");
+                    resource.failed();
+                    return;
+                }
+
+                let mut gamma = Vec::new();
+                for i in 0..buf.len() / 2 {
+                    gamma.push(((buf[i * 2 + 1] as u16) << 8) | buf[i * 2] as u16);
+                }
+                if data.gamma_size.is_none() {
+                    error!("gamma size not set");
+                    resource.failed();
```

Comment:
We already failed it when it was created in this case, probably can just `return` silently.

Or, even better, add a flag `data.failed`, set it when failing the resource, and just do `if data.failed { return; }` at the top.